server/src/actors/room_manager_actor.rs

```diff
@@ -0,0 +1,152 @@
+use crate::actors::GameActor;
+use actix::prelude::*;
+use rand::{distributions::Alphanumeric, Rng};
+use std::{
+    collections::HashMap,
+    io::{Error, ErrorKind, Result},
+    sync::Mutex,
+};
+use tokyo::models::GameConfig;
+
+pub struct RoomManagerActor {
+    config: GameConfig,
+    id_counter: Mutex<u32>,
+    rooms: Mutex<HashMap<String, Room>>,
+}
+
+struct Room {
+    id: String,
+    name: String,
+    max_players: u32,
+    time_limit_seconds: u32,
+    token: String,
+    game: Addr<GameActor>,
+}
+
+impl Room {
+    pub fn new(
+        config: &GameConfig,
+        id: String,
+        name: String,
+        max_players: u32,
+        time_limit_seconds: u32,
+        token: String,
+    ) -> Room {
+        let game_actor = GameActor::new(config.clone());
+        let game_actor_addr = game_actor.start();
+        Room { id, name, max_players, time_limit_seconds, token, game: game_actor_addr }
+    }
+}
+
+impl RoomManagerActor {
+    pub fn new(cfg: GameConfig) -> RoomManagerActor {
+        RoomManagerActor {
+            config: cfg,
+            id_counter: Mutex::new(0),
+            rooms: Mutex::new(HashMap::new()),
```

Comment:
RwLock is better than Mutex, since mostly `rooms` will be read (on join & action) instead of write (add room/edit room name).

Reply:
Actually, we will keep the idea from the beginning. When a user connect to game server using `room_token` or `room_id` we will make a connect with the Addr<GameActor>. How do u think it can affect in this case?
```rust
pub fn socket_handler(
    (req, state, query): (HttpRequest<AppState>, State<AppState>, Query<QueryString>),
) -> Result<actix_web::HttpResponse, actix_web::Error> {
    if crate::APP_CONFIG.dev_mode || crate::APP_CONFIG.api_keys.contains(&query.key) {
        let r = state
            .room_manager_addr
            .send(JoinRoom {
                room_token: query.room_token.clone(),
                api_key: query.key.clone(),
                team_name: query.name.clone(),
            })
            .wait()
            .unwrap();
        match r {
            Ok(room) => actix_web::ws::start(
                &req,
                ClientWsActor::new(room.game_addr, query.key.clone(), query.name.clone()),
            ),
            Err(err) => Err(actix_web::error::ErrorBadRequest(err.to_string())),
        }
    } else {
        Err(actix_web::error::ErrorBadRequest("Invalid API Key"))
    }
}
```

Reply:
ah yeah, I forgot ClientWsActor only need to join room once. Yes, this is great.